test/hotspot/jtreg/gc/TestReferenceRefersToDuringConcMark.java

```diff
@@ -0,0 +1,206 @@
+    private static volatile TestObject testObject = null;
+
+    private static ReferenceQueue<TestObject> queue = null;
+
+    private static WeakReference<TestObject> testWeak = null;
+
+    private static void setup() {
+        testObjectNone = new TestObject(0);
+        testObject = new TestObject(4);
+
+        queue = new ReferenceQueue<TestObject>();
+
+        testWeak = new WeakReference<TestObject>(testObject, queue);
+    }
+
+    private static void gcUntilOld(Object o) throws Exception {
+        if (!WB.isObjectInOldGen(o)) {
+            WB.fullGC();
+            if (!WB.isObjectInOldGen(o)) {
+                fail("object not promoted by full gc");
+            }
+        }
+    }
+
+    private static void gcUntilOld() throws Exception {
+        gcUntilOld(testObjectNone);
+        gcUntilOld(testObject);
+
+        gcUntilOld(testWeak);
+    }
+
+    private static void progress(String msg) {
+        System.out.println(msg);
+    }
+
+    private static void fail(String msg) throws Exception {
+        throw new RuntimeException(msg);
+    }
+
+    private static void expectCleared(Reference<TestObject> ref,
+                                      String which) throws Exception {
+        expectNotValue(ref, testObjectNone, which);
+        if (!ref.refersTo(null)) {
+            fail("expected " + which + " to be cleared");
+        }
+    }
+
+    private static void expectNotCleared(Reference<TestObject> ref,
+                                         String which) throws Exception {
+        expectNotValue(ref, testObjectNone, which);
+        if (ref.refersTo(null)) {
+            fail("expected " + which + " to not be cleared");
+        }
+    }
+
+    private static void expectValue(Reference<TestObject> ref,
+                                    TestObject value,
+                                    String which) throws Exception {
+        expectNotValue(ref, testObjectNone, which);
+        expectNotCleared(ref, which);
+        if (!ref.refersTo(value)) {
+            fail(which + " doesn't refer to expected value");
+        }
+    }
+
+    private static void expectNotValue(Reference<TestObject> ref,
+                                       TestObject value,
+                                       String which) throws Exception {
+        if (ref.refersTo(value)) {
+            fail(which + " refers to unexpected value");
+        }
+    }
+
+    private static void checkInitialStates() throws Exception {
+        expectValue(testWeak, testObject, "testWeak");
+    }
+
+    private static void discardStrongReferences() {
+        testObject = null;
+    }
+
+    private static void testConcurrentCollection() throws Exception {
+        progress("setup concurrent collection test");
+        setup();
+        progress("gcUntilOld");
+        gcUntilOld();
+
+        progress("acquire control of concurrent cycles");
+        WB.concurrentGCAcquireControl();
```

Comment:
I think this test could be made a lot smaller and more obvious if it was explicitly just testing the keep-alive behavior of Reference.get for most concurrent collectors, rather than being a trimmed down copy of the earlier test.

Reply:
Right. I trimmed it some more. I think it's cleaner now.